ngraph/core/include/openvino/pass/graph_rewrite.hpp

```diff
@@ -0,0 +1,249 @@
+    /// All registered transformations will be executed in a single graph traversal.
+    /// Example below show the basic usage of pass::GraphRewrite
+    ///
+    ///     pass::Manager manager;
+    ///     auto anchor = manager.register_pass<GraphRewrite>();
+    ///     anchor->add_matcher<MatcherPassA>();
+    ///     anchor->add_matcher<MatcherPassB>();
+    ///     anchor->set_name("CommonMatchers");
+    ///     manager.run_passes(f);
+    ///
+    /// For some purposes transformation can be registered and disabled by default.
+    ///
+    ///     anchor->add_matcher<MatcherPassB, false>();
+    ///
+    /// \return shared_ptr to the transformation instance
+    template <typename T,
+              bool Enabled = true,
+              class... Args,
+              typename std::enable_if<std::is_base_of<pass::MatcherPass, T>::value, bool>::type = true>
+    std::shared_ptr<T> add_matcher(Args&&... args) {
+        static_assert(std::is_base_of<pass::MatcherPass, T>::value, "pass not derived from MatcherPass");
+        auto pass = std::make_shared<T>(std::forward<Args>(args)...);
+        auto pass_config = get_pass_config();
+        pass->set_pass_config(pass_config);
+        if (!Enabled && !pass_config->is_enabled<T>()) {
+            pass_config->disable<T>();
+        }
+        m_matchers.push_back(pass);
+        return pass;
+    }
+
+    /// \brief Register passes from GraphRewrite class that contains sequence of matcher
+    /// passes registered in its ctor.
+    /// For example:
+    ///
+    ///    class ov::pass::LinFusions: public ov::pass::GraphRewrite {
+    ///    public:
+    ///         OPENVINO_RTTI_DECLARATION;
+    ///         Fusions() {
+    ///             add_matcher<ov::pass::AddFusion>();
+    ///             add_matcher<ov::pass::MulFusion>();
+    ///         }
+    ///     };
+    ///
+    ///     pass::Manager manager;
+    ///     auto anchor = manager.register_pass<GraphRewrite>();
+    ///     anchor->add_matcher<LinFusions>();
+    ///     anchor->add_matcher<OtherFusions>();
+    ///     anchor->set_name("CommonFusions");
+    ///     manager.run_passes(f);
+    ///
+    /// In this case all matcher passes from LinFusions pass will be united with other
+    /// registered matchers.
+    template <typename T,
+              class... Args,
+              typename std::enable_if<std::is_base_of<pass::GraphRewrite, T>::value, bool>::type = true>
+    void add_matcher(Args&&... args) {
+        static_assert(std::is_base_of<pass::GraphRewrite, T>::value, "pass not derived from GraphRewrite");
+        auto pass = std::make_shared<T>(std::forward<Args>(args)...);
+        auto pass_config = get_pass_config();
+
+        for (auto& matcher : pass->m_matchers) {
+            pass->set_pass_config(pass_config);
+            m_matchers.push_back(matcher);
+        }
+    }
+
+    OPENVINO_DEPRECATED("Use MatcherPass instead")
+    void add_matcher(const std::shared_ptr<pattern::Matcher>& m,
```

Comment:
can we remove it finally?

Reply:
It can be backward incompatible for our old users.

Reply:
why? it's deprecated for several releases

Reply:
Hmm, yes, it is a valid comment. So I propose to remove these deprecated methods in the separate PR.

Reply:
ok, let's merge then

Reply:
It is still used by some OV components actually